### PR TITLE
Shader API improvements

### LINKED ISSUE #546: No replacement option for Geometry Shaders
The parameter `shader_type` in the function `shader_to_actor` includes the option `geometry` as seen here:

https://github.com/fury-gl/fury/blob/dadbb8e66938e964c25d01e6bed427e1fe7efc0c/fury/shaders/base.py#L64

However, attempting to replace a geometry shader does not produce any effect. This has been already spotted in VTK's discourse ([link](https://discourse.vtk.org/t/geometry-shader-implementation/1189/3)).

### PR DESCRIPTION
**Pull request by guaje**
This PR includes the following improvements:

 - Improved docstrings for the shaders API.
 - Increased test coverage for the shaders functions.
   - Removed duplicated tests.
 - Fixed #546.
 - Fixed billboard actor (Only actor affected by the described changes).

### CODE REVIEW

**Comment by skoudoro** [comment]
Hi @guaje,

Why this PR is a draft? Do you plan to add more stuff in here?

**Comment by guaje** [comment]
> Why this PR is a draft? Do you plan to add more stuff in here?

Yes, I'm planning to move the `get_code` function out of the billboard actor. And add some tests to that and the `replace_shader_in_actor` function.

File: fury/shaders/base.py [hunk 1]
```diff
@@ -4,10 +4,16 @@
 from fury.lib import (numpy_support, Command, VTK_OBJECT,
                       calldata_type, DataObject, Shader)
 
-SHADERS_TYPE = {"vertex": Shader.Vertex,
-                "geometry": Shader.Geometry,
-                "fragment": Shader.Fragment,
-                }
+from os.path import join as pjoin, dirname, isfile
```
Comment (skoudoro) [inline fury/shaders/base.py:7]: pep8:  import order.   see https://peps.python.org/pep-0008/#imports for more information

File: fury/shaders/tests/test_base.py [hunk 2]
```diff
@@ -1,11 +1,17 @@
-
 import numpy as np
 import numpy.testing as npt
+import os
+import pytest
```
Comment (skoudoro) [inline fury/shaders/tests/test_base.py:4]: pep8: import order, see above

File: fury/shaders/tests/test_base.py [hunk 3]
```diff
@@ -88,6 +140,15 @@
     """
 
 
+def generate_cone():
+    cone_src = ConeSource()
+    cone_mapper = PolyDataMapper()
+    cone_mapper.SetInputConnection(cone_src.GetOutputPort())
+    cone_actor = Actor()
+    cone_actor.SetMapper(cone_mapper)
+    return cone_actor
```
Comment (skoudoro) [inline fury/shaders/tests/test_base.py:149]: Why restart from scratch and not use an actor?   

To reduce the maintenance, you should try to reduce as much as possible code duplication

File: fury/shaders/tests/test_base.py [hunk 4]
```diff
@@ -97,26 +158,37 @@ def generate_cube_with_effect():
...
-                      vertex_impl)
+    set_polydata_colors(polydata, colors)
+
+    mapper = PolyDataMapper()
+    mapper.SetInputData(polydata)
+    mapper.SetVBOShiftScaleMethod(False)
+
+    point_actor = Actor()
+    point_actor.SetMapper(mapper)
+
```
Comment (skoudoro) [inline fury/shaders/tests/test_base.py:190]: Why restart from scratch and not use a point actor?   

To reduce the maintenance, you should try to reduce as much as possible code duplication

File: fury/shaders/tests/test_base.py [hunk 5]
```diff
@@ -153,6 +225,60 @@ def my_cbk(_caller, _event, calldata=None):
     report = window.analyze_snapshot(arr)
     npt.assert_equal(report.objects, 1)
 
+    cone_actor = generate_cone()
```
Comment (skoudoro) [inline fury/shaders/tests/test_base.py:228]: use actor instead of a new custom one

File: fury/shaders/tests/test_base.py [hunk 6]
```diff
@@ -162,3 +288,100 @@ def test_attribute_to_actor():
...
+    code = get_shader_code(list_str)
+    npt.assert_equal(code, '\n' + '\n'.join(list_str))
+    # Test list of str fname
+    code = get_shader_code(list_fname)
+    npt.assert_equal(code, '\n' + '\n'.join(list_str))
+    # Test list of str code and str fname
+    code = get_shader_code(list_mix)
+    npt.assert_equal(code, '\n' + '\n'.join(list_str))
+    os.remove(pname_test1)
+    os.remove(pname_test2)
```
Comment (skoudoro) [inline fury/shaders/tests/test_base.py:323]: use TemporaryFolder here

File: fury/shaders/tests/test_base.py [hunk 7]
```diff
@@ -162,3 +288,100 @@ def test_attribute_to_actor():
...
+    dummy_file_name = 'dummy.txt'
+    dummy_file_contents = 'This is some dummy text.'
+
+    dummy_file = open(os.path.join(SHADERS_DIR, dummy_file_name), 'w')
+    dummy_file.write(dummy_file_contents)
+    dummy_file.close()
+
+    npt.assert_string_equal(load(dummy_file_name), dummy_file_contents)
+
+    os.remove(os.path.join(SHADERS_DIR, dummy_file_name))
```
Comment (skoudoro) [inline fury/shaders/tests/test_base.py:336]: same as above.  Use `with` and use `temporaryfolder`

**Review by skoudoro — COMMENTED** [review]
Hi @guaje,

Thank you for this update. See below for my comments

File: fury/shaders/base.py (hunk 1 above)
Comment (guaje) [reply fury/shaders/base.py]: This should be fixed, but PTAL.

File: fury/shaders/tests/test_base.py (hunk 6 above)
Comment (guaje) [reply fury/shaders/tests/test_base.py]: Fixed in latest commits.

File: fury/shaders/tests/test_base.py (hunk 7 above)
Comment (guaje) [reply fury/shaders/tests/test_base.py:360]: This test was moved from `test_shaders.py` to `test_base.py`. It looks for text files in the fury/shaders directory so idk if the `TemporaryDirectory` will be the best fit for this.

File: fury/shaders/tests/test_base.py (hunk 5 above)
Comment (guaje) [reply fury/shaders/tests/test_base.py]: Done.

File: fury/shaders/tests/test_base.py (hunk 3 above)
Comment (guaje) [reply fury/shaders/tests/test_base.py]: The test using this function was moved from `test_shader.py` to `test_base.py`. I have updated it by using `actor.cone` instead.

**Review by skoudoro — APPROVED** [review]
Hi @guaje,

thank you for the update. merging

**Pull request merged**